### PR TITLE
Implemented writing an xmp sidecar file on export

### PR DESCRIPTION
**Pull request by bertwim**
This contribution adds in the preferences/storage/xmp section the option to write an xmp sidecar file "on export". The option is almost identical to the case where an xmp file is never written, except when an export of an image is explicitly requested. With this option, one can always be assured that the xmp is consistent with the exported image. It fits a workflow where the work on a (raw) image is considered done only when an export (say of a jpg file) has been done. And there is no accidental generation of xmp-files for images that have not been edited yet. 

The impact on the code is in three places:

-   in data/darktableconfig.xml.in the option "on export" is added to the appropriate combo (including tooltip text)
-   In common/image.c the (string)value of the new option is handled, in the routine dt_image_get_xmp_mode(). . Note that it is not required to introduce another enum value. The new option corresponds to the existing value DT_WRITE_XMP_NEVER. In addition, a (new) boolean configuration variable "write_sidecar_on_export" is being used: its value defaults to FALSE, but set to TRUE when the user has selected the "on export" option.
-   In libs/export.c, routine _export_button_clicked(), an if-statement is added to write the xmp files when the mentioned boolean variable has been set to TRUE.

### CODE REVIEW

**Comment by TurboGit** [comment]
You know that the xmp can be embedded into the jpeg?

> With this option, one can always be assured that the xmp is consistent with the exported image.

That's also the case when the xmp is always written. Also if you export you're image multiple times (different format, size...) you'll have a single XMP possibly not consistent with all exported images.

At the end I'm not sure the extra complexity is worth it.

**Comment by bertwim** [comment]
> 

The point is that I don't want xmp files be *always* written, but only for images that I have been editing. This keeps my directory clean. The camera produces both raw+jpg. With the option to always generate an xmp files, one gets also xmp files for images that have not, or not yet, been worked on with DT.  I find this not only annoying but also confusing. I think it is absolutely worth it. The code impact is minimal and it gives a lot of clarity.

**Comment by TurboGit** [comment]
> The point is that I don't want xmp files be always written, but only for images that I have been editing.

There is already an option for that.

**Comment by bertwim** [comment]
> 
Are you referring to the after edit option? That is not precisely the same. It writes after an edit, but  then introduces an inconsistency with the available jpg. This is want I want to prevent.

**Comment by bertwim** [comment]
> 

I don't agree. It think it is very equivalent, and certainly an enrichment,  to the existing options of "never," "always", "after edit", "on import".  Sure, as always, there will be other possibilities to achieve the same using scripting, but that is not equivalent when it comes to user experience.

**Comment by parafin** [comment]
XMP is a backup for data stored in the darktable DB. All these options just define when this backup is first generated, but after that it's kept in sync with DB. Your proposal is very different in that regard and breaks that sync and meaning of XMP files. It's rather bound to exported JPG file (which BTW usually already contains the XMP inside it), that's why I'm suggesting to put it alongside export, not imported raw file.

**Comment by parafin** [comment]
What you're describing sounds like a bug with "after edit" option. Maybe it happens because you have some auto-applied presets. I don't know exactly how it's supposed to work.

Merging the feature from this PR is not the right way to fix this problem for sure.

**Comment by TurboGit** [comment]
> Maybe it happens because you have some auto-applied presets. I don't know exactly how it's supposed to work.

It should not be the case, the XMP is only written when entering the darkroom IIRC. So look like a bug indeed.

**Comment by bertwim** [comment]
> 
Yes, indeed. I spend a good part of the afternoon in trying to understand DTs behaviour with respect to the option "after edit". Something is really wrong. Sometimes xmp files are written with this option set, sometimes not...  I do not yet see the consistency.   The "never" option is consistently ok, but that is not surprising (I think).

**Comment by gi-man** [comment]
> > 
> 
> Yes, indeed. I spend a good part of the afternoon in trying to understand DTs behaviour with respect to the option "after edit". Something is really wrong. Sometimes xmp files are written with this option set, sometimes not... I do not yet see the consistency. The "never" option is consistently ok, but that is not surprising (I think).

Are you applying some rating or tags on import?

**Comment by bertwim** [comment]
> Technically the ˋafter editˋ is checked by comparing the hash against the one generated immediately after import and all presets applied. Don't know atm how the raw+jpeg Situation might interfere here.

I find it  hard to understand how this mechanism of deciding if writing an xmp is required works. What is this hashing? But when I grep for DT_WRITE_XMP in the code there are  occurences (e.g. in dt_image_write_sidecar_files) where the test is essentially 
xmp_mode != DT_WRITE_XMP_NEVER. In the latter routine there is a comment suggesting this is not good enough....

**Comment by jenshannoschwalm** [comment]
I did that code some time ago, it was somewhat hard to do but you will find the relevant changes in #9681

I am somewhat short of time this week, my guess is that a jpeg plus raw file does some grouping or alike not reflected by the checked hash status.

**Comment by bertwim** [comment]
I have done some further experimentation to see when the xmp are automatically being generated
when the xmp-write option "after edit" is active.

This is what I did.
- In a (test)-directory with both raw (.nef) and jpg-files (from camera) I started darktable under
  different pre-conditions. In all cases I first ensured no xmp files are present in the test directory.
  Darktable was started from the command line, with the test director as commandline argument: "darktable .".
- experiment 1: cache removed, database files in .config directory deleted.
                result: NO XMP FILES generated
- experiment 2: only restart DT again and quit. Nothing else.
                result: xmp files generated (in my case, 19 in total)
- experiment 3: remove xmp and the cache.
                result: xmp files generated again.
- experiment 4: remove xmp, the cache and the data.db file in .config
                result: xmp files generated again.
- experiment 5: remove xmp, the cache and the library.db file in .config
                result: NO XMP FILES are being generated.
                
Furthermore, when I remove the jpg files, the results are exactly the same as with these files present.

I put the testing in a script, the output of which is quite readable, I think
After establishing the conditions mentioned above, darktable is being run and
output is generated as follows:
```
darktable .
echo "#xmp="`ls *.* | grep xmp | wc -l`
ls -l ~/.config/darktable/*.db
```
That is, it echoes the number of xmp files + lists the db files in the config directory.

Now follows the output when all experiments are run:

```
+++ experiment 1 ++++++++++++++++++
rm -rf *.xmp
rm -rf ~/.cache/darktable/*
rm -rf ~/.config/darktable/data*
rm -rf ~/.config/darktable/library*

source ./results
#xmp=0
-rw-r--r-- 1 bertwim users  307200 dec  9 12:56 /home/bertwim/.config/darktable/data.db
-rw-r--r-- 1 bertwim users 1310720 dec  9 12:56 /home/bertwim/.config/darktable/library.db
==================================
+++ experiment 2 +++++++++++++++++
source ./results
#xmp=19
-rw-r--r-- 1 bertwim users  307200 dec  9 12:56 /home/bertwim/.config/darktable/data.db
-rw-r--r-- 1 bertwim users 1310720 dec  9 12:56 /home/bertwim/.config/darktable/library.db
==================================
+++ experiment 3 +++++++++++++++++
rm -rf *.xmp
rm -rf ~/.cache/darktable/*

source ./results
#xmp=19
-rw-r--r-- 1 bertwim users  307200 dec  9 12:56 /home/bertwim/.config/darktable/data.db
-rw-r--r-- 1 bertwim users 1376256 dec  9 12:56 /home/bertwim/.config/darktable/library.db
==================================
+++ experiment 4 +++++++++++++++++
rm -rf *.xmp
rm -rf ~/.cache/darktable/*
rm -rf ~/.config/darktable/data.db

source ./results
#xmp=19
-rw-r--r-- 1 bertwim users  307200 dec  9 12:57 /home/bertwim/.config/darktable/data.db
-rw-r--r-- 1 bertwim users 1376256 dec  9 12:57 /home/bertwim/.config/darktable/library.db
==================================
+++ experiment 5 +++++++++++++++++
rm -rf *.xmp
rm -rf ~/.cache/darktable/*
rm -rf ~/.config/darktable/library.db

source ./results
#xmp=0
-rw-r--r-- 1 bertwim users  307200 dec  9 12:57 /home/bertwim/.config/darktable/data.db
-rw-r--r-- 1 bertwim users 1310720 dec  9 12:57 /home/bertwim/.config/darktable/library.db
==================================

```

**Comment by bertwim** [comment]
> > Maybe it happens because you have some auto-applied presets. I don't know exactly how it's supposed to work.
> 
> It should not be the case, the XMP is only written when entering the darkroom IIRC. So look like a bug indeed.

i would like to re-open the discussion, as it seems that tacitly this thread has been closed by marking it "Controversial". Why controversial?
The discussion went into the direction of "you are trying to add a new feature X, but that is only because feature Y is buggy. Yes, the option "after edit" is buggy. But that is not the reason why the option of "xmp on export" is needed. I think the proposed feature is independent of the "on edit" option, the latter being buggy or not.  To me, the desire that the exported xmp file is in sync with the exported jpg file is an absolutely valid user requirement, which can be achieved by a very minor code adjustment. Moreover, I fail to see how this option could jeopardize the alleged DT-database backup-strategy in any way.  Although alleged, this has not been explained at all.

**Comment by parafin** [comment]
XMP files that lie near imported to the library files are not “exported”, they mirror the state of imported file. Logically “exported” XMP files should be placed near the exported files, otherwise it gets very confusing. But exported JPG files by default already have XMP embedded in them, plus it’s probably possible to implement this via lua script.

And you actually haven’t explained the use-case for this feature if the mentioned bug is fixed. So I feel like I’m repeating myself, probably because you haven’t provided any new information, nor do I see anyone else advocating for this feature.

**Comment by bertwim** [comment]
> XMP files that lie near imported to the library files are not “exported”, they mirror the state of imported file. Logically “exported” XMP files should be placed near the exported files, otherwise it gets very confusing. But exported JPG files by default already have XMP embedded in them, plus it’s probably possible to implement this via lua script.
> 
> And you actually haven’t explained the use-case for this feature if the mentioned bug is fixed. So I feel like I’m repeating myself, probably because you haven’t provided any new information, nor do I see anyone else advocating for this feature.

Well, indeed I wonder why this issue has not come up earlier! The common use case where you have a directory full of pictures (raw+jpg), but where you want to edit only some of these is not well supported by DT. Generating default xmp files *on import* of a picture is not very meaningful when you don't intend to modify the image at all!  On the contrary, these xmp files are misleading. But we probably agree that they shouldn't be there in that case. But suppose the option "after edit"  is working fine. It means that we then have at all times an xmp file that represents the state of the edited image. This is fine, nothing wrong with that if you choose for it. But for me, an image is ready when I have exported it. That is the situation I want to keep. Even when the exported jpg has to be regenerated from the xmp. If I (accidentally/experimentally) change some DT settings which I do not approve. I do not want to see those reflected in a changed xmp file. I want an xmp file I know that is my choice. Even when the jpg is gone! Now this is a user requirement. Other people will feel different and that is also fine. This is my use case.

**Comment by jenshannoschwalm** [comment]
I checked and indeed the test for being altered (edited) is not correct for jpegs. Will test and investigate the next days and we will see any remaining issue.

**Comment by parafin** [comment]
Why would you want XMP file be de-synchronised with what you see inside darktable? It’s very unintuitive situation, that’s why I’m against it (plus I’m against adding more options to the settings, it’s complicated enough already). What do you do with those XMP files?

**Comment by bertwim** [comment]
> Why would you want XMP file be de-synchronised with what you see inside darktable? It’s very unintuitive situation, that’s why I’m against it (plus I’m against adding more options to the settings, it’s complicated enough already). What do you do with those XMP files?

For me it is quite simple: I see the xmp as the "recipe" of how to get from the original raw file to the jpg-file that I ultimately want to have and keep.  It contains all the modifications on the raw file that I value and hence want to keep. It is the file I need if I want to regenerate the jpg if needed (think of automated workflows!). It is the file that matters.
I do not want such a file to be sensitive to any accidental/experimental changes I happen to have made with DT.  That is, I do not care if the xmp does not store the latest -but not necessarily the greatest- state of darktable. It shouldn't!.  I want it to store the state I intended. The moment I pressed "export".  Nothing else.

**Comment by gi-man** [comment]
Bertwim:

You can do that already. When you export the jpg, if you include the develop history (the recipe) to the jpg (see export preferences), you can load the jpg as the load sidecar file (history stack module). 

_OR_

You can also use the Copy file format when you export. This will make a copy of the Raw + xmp. Just send them to a "finished" folder.

**Comment by parafin** [comment]
> > Why would you want XMP file be de-synchronised with what you see inside darktable? It’s very unintuitive situation, that’s why I’m against it (plus I’m against adding more options to the settings, it’s complicated enough already). What do you do with those XMP files?
> 
> For me it is quite simple: I see the xmp as the "recipe" of how to get from the original raw file to the jpg-file that I ultimately want to have and keep. It contains all the modifications on the raw file that I value and hence want to keep. It is the file I need if I want to regenerate the jpg if needed (think of automated workflows!). It is the file that matters. I do not want such a file to be sensitive to any accidental/experimental changes I happen to have made with DT. That is, I do not care if the xmp does not store the latest -but not necessarily the greatest- state of darktable. It shouldn't!. I want it to store the state I intended. The moment I pressed "export". Nothing else.

You’re not actually answering my question. What is the workflow that requires XMP files to be managed like this? Is this about backups? How do you restore the state inside darktable to the “exported” XMP state? I don’t see a straightforward way, which makes whole idea flawed. Basically what you want is to make a duplicate after you’re satisfied with the result, so that it could be left untouched while you can continue to experiment with the original. Maybe having a read-only duplicate would be a good feature request, and again it could all (export + duplicate) be automated with lua (which you totally ignore, even though you mention automated workflows yourself).

Basically what I’m saying is that your idea as in this PR conflicts with some core principles or expectations of darktable, and so is a bad way to meet your requirements. I suggest you give alternative suggestions as mentioned in this PR by me and others a go, instead of pushing this feature as-is.

**Comment by jenshannoschwalm** [comment]
I tested the behaviour of "after edit" quite hard yesterday and can not confirm any bug so far, the testing for "image is altered" works as intended.

There are a few peculiar things though

1. If a jpeg file is grouped to a raw, this grouping marks it as edited. So if I remove the image from the database there is an xmp written even if un-edited.

2. Deleting the history always forces to write an xmp on all images. 

We might want to do something on these but it's not an easy hit by reading the source. So would it be worth it? 
I would "fix" that if desired...

**Comment by bertwim** [comment]
> I tested the behaviour of "after edit" quite hard yesterday and can not confirm any bug so far, the testing for "image is altered" works as intended.
> 
> There are a few peculiar things though
> 
>     1. If a jpeg file is grouped to a raw, this grouping marks it as edited. So if I remove the image from the database there is an xmp written even if un-edited.
> 
>     2. Deleting the history always forces to write an xmp on all images.
> 
> 
> We might want to do something on these but it's not an easy hit by reading the source. So would it be worth it? I would "fix" that if desired...

Not sure what you mean by 1., but I think issue 2. is consistent with my series of observations described above. 
To me, the generation of an xmp-file for any image that I have not edited is a no go. Spurious xmp-files destroy my workflow which contains rsyncs of image files (+associated xmp, if present) to mirror locations. I think it is a conceptual flaw to generate an xmp-file before it is even known that the associated images (raw+jpg) will be edited, a flaw which in my view needs to be ironed out. 
To understand when xmp-files are written I found the associated code  indeed hard to find, to read and understand the associated code.  
As I have explained many times now, my contribution is not intended to fix the issue mentioned here -I wouldn't even know how to do it-, but it gives a simple way forward where the problem does not occur and where I also fail to see how it would jeopardize DTs philosophy. I am using this approach for several years now with any issues. Which made me think that it would be beneficial for others as well...

**Comment by bertwim** [comment]
> @bertwim did you notice late commits fixing the xmp writing issue for jpegs?

I hadn't noticed it, as I was tired of the fruitless and useless discussion about the usage of xmp-files. But thanks for letting me know. I will check if my issues can be overcome!
